### DISCUSSION TITLE
VS Code extension secrets when default keyring is unavailable/locked

### DISCUSSION THREAD

**Discussion by cgrama** [microsoft/vscode-discussions#1188] [Extension Development QnA] [answered]
When default keyring is unavailable or locked, extensions are able to store/get/delete extension secrets using VSCode secret storage apis with NO errors. 
We get a vscode notification (in the notification windos) "You're running in a GNOME environment but the OS keyring is not available for encryption. Ensure you have gnome-keyring or another libsecret compatible implementation installed and running.".
There is NO error returned warning that secrets are NOT stored on the disk.
The extension's secrets stored in this scenarios are available for use by the extension in the same vscode session. If you close and reopen vscode the extension's secrets are no longer available.
This makes it very difficult for secrets to be used by vscode extensions as the extension does not know if the secret was stored in memory or on the disk.
Questions:
(1) How can a vscode extension check if the extension secret is stored in-memory or in the vscode secret storage ?
(2) Can the extension secret store apis be enhanced to return an error or return value which informs the api caller that the secret is stored in-memory and not in vscode secret storage ?
(3) Also from the secret storage info shared at https://github.com/microsoft/vscode-discussions/discussions/748, is the default keyring used only for encryption OR is the default keyring used for storing the extension secrets also ? 

Included below is a prt of the vscode console output from "code --user-data-dir=/scratch/vscode2 code --verbose --vmodule="*/components/os_crypt/*=1" | tee verbose_upgrade.log":
app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003951.673893:INFO:CONSOLE(656)] "%cTRACE color: #888 [mainThreadSecretState] Setting password for vscextn.extn1 extension:  key1", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003951.675050:VERBOSE1:key_storage_util_linux.cc(52)] Password storage detected desktop environment: GNOME
[2669207:0510/003951.675065:VERBOSE1:key_storage_linux.cc(118)] Selected backend for OSCrypt: GNOME_LIBSECRET
[2669207:0510/003954.651823:VERBOSE1:libsecret_util_linux.cc(169)] Dummy store to unlock the default keyring failed: No such interface “org.freedesktop.Secret.Collection” on object at path /org/freedesktop/secrets/collection/Default_5fkeyring
[2669207:0510/003954.651849:VERBOSE1:key_storage_linux.cc(165)] OSCrypt using Libsecret as backend.
[2669207:0510/003957.531557:VERBOSE1:key_storage_libsecret.cc(73)] Libsecret lookup failed: No such interface “org.freedesktop.Secret.Collection” on object at path /org/freedesktop/secrets/collection/Default_5fkeyring
[2669207:0510/003957.538272:INFO:CONSOLE(656)] "%cTRACE color: #888 [SecretStorageService] Encryption is not available, falling back to in-memory storage", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.545398:INFO:CONSOLE(656)] "%cTRACE color: #888 [NativeSecretStorageService] Notifying user that secrets are not being stored on disk.", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.548906:INFO:CONSOLE(796)] "You're running in a GNOME environment but the OS keyring is not available for encryption. Ensure you have gnome-keyring or another libsecret compatible implementation installed and running.", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (796)
[2669207:0510/003957.549118:INFO:CONSOLE(656)] "%cTRACE color: #888 [secrets] encrypting secret for key: {"extensionId":"vscextn.extn1","key":key1}", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.549318:INFO:CONSOLE(656)] "%cTRACE color: #888 [secrets] storing encrypted secret for key: secret://{"extensionId":"vscextn.extn1","key":key1}", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.549540:INFO:CONSOLE(656)] "%cTRACE color: #888 [SecretStorageService] Notifying change in value for secret: {"extensionId":"vscextn.extn1","key":key1}", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.552598:INFO:CONSOLE(656)] "%cTRACE color: #888 [secrets] stored encrypted secret for key: secret://{"extensionId":"vscextn.extn1","key":key1}", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.558749:INFO:CONSOLE(656)] "%cTRACE color: #888 [mainThreadSecretState] Password set for:  vscextn.extn1 key1", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.636472:INFO:CONSOLE(656)] "%cTRACE color: #888 [mainThreadSecretState] Getting password for vscextn.extn1 extension:  key1", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.638363:INFO:CONSOLE(656)] "%cTRACE color: #888 [secrets] getting secret for key: secret://{"extensionId":"vscextn.extn1","key":key1}", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.638899:INFO:CONSOLE(656)] "%cTRACE color: #888 [secrets] decrypting gotten secret for key: secret://{"extensionId":"vscextn.extn1","key":key1}", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.641229:INFO:CONSOLE(656)] "%cTRACE color: #888 [secrets] decrypted secret for key: secret://{"extensionId":"vscextn.extn1","key":key1}", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.641454:INFO:CONSOLE(656)] "%cTRACE color: #888 [mainThreadSecretState] Password found for:  vscextn.extn1 key1", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.648246:INFO:CONSOLE(656)] "%cTRACE color: #888 [mainThreadSecretState] Getting password for vscextn.extn1 extension:  key1", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.648418:INFO:CONSOLE(656)] "%cTRACE color: #888 [secrets] getting secret for key: secret://{"extensionId":"vscextn.extn1","key":key1}", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.648558:INFO:CONSOLE(656)] "%cTRACE color: #888 [secrets] decrypting gotten secret for key: secret://{"extensionId":"vscextn.extn1","key":key1}", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.649588:INFO:CONSOLE(656)] "%cTRACE color: #888 [secrets] decrypted secret for key: secret://{"extensionId":"vscextn.extn1","key":key1}", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)
[2669207:0510/003957.650248:INFO:CONSOLE(656)] "%cTRACE color: #888 [mainThreadSecretState] Password found for:  vscextn.extn1 key1", source: vscode-file://vscode-app/usr/share/code/resources/app/out/vs/workbench/workbench.desktop.main.js (656)

**Comment by TylerLeonhardt** [comment]
Hi @cgrama, I have some context and a followup question:

> (1) How can a vscode extension check if the extension secret is stored in-memory or in the vscode secret storage ?
> (2) Can the extension secret store apis be enhanced to return an error or return value which informs the api caller that the secret is stored in-memory and not in vscode secret storage ?

As you've called out, we do notify the user that their keyring is in a bad state and they need to remediate. Including attaching our troubleshooting guide. It's intended to call out that they are in a bad state... Any solution would either be error prone (us fixing it for them) or a questionable security practice (storing the secrets in plaintext).

With that said, my question for you is... what would you like to do with this information that they are in a bad state?

> (3) Also from the secret storage info shared at https://github.com/microsoft/vscode-discussions/discussions/748, is the default keyring used only for encryption OR is the default keyring used for storing the extension secrets also ?

The default keyring is only used for encryption. In other words, it holds a piece of the encryption key needed to encrypt/decrypt secrets which are then stored in the user's data directory with other state like UI state.

**Reply by cgrama** [reply]
Hi @TylerLeonhardt,
When the vscode extension tries to store an extension secret using the vscode secret storage "store" api,  the vscode user gets a notification regarding the bad state. But the vscode extension does not know about the notification shown for the bad state.
The secret storage "store" api  used by the vscode extension does not return any error or exception.
In the same vscode session, the vscode extension tries to get the secret using the secret storage "get" api and that is successful as well.
So from the vscode extension perspective, the extension secret has been saved successfully to the vscode secret storage.
When user closes the vscode session and reopens vscode,  the vscode extension will fail to get the secret, as the secret is no longer available and the extension will notify/prompt the user for the missing secret.
If the secret storage "store" api would return an error or throw an exception then the vscode extension will know that something went wrong when saving to the vscode secret storage.
The vscode secret storage "store" api should notify the caller of the api about the bad state either by returning an error or throwing an exception in addition to the errors that popup in the vscode notification window. The vscode extension can decide to take some action based on this error/exception to notify the vscode user that secrets cannot be saved to vscode secret storage and refer to the notification for more detailed info about the keyring issues. 
Also, In vscode 1.89.1 and earlier versions, is there any way for a vscode extension using secret storage "store" api to check if the secret was saved using in-memory store or the vscode secret storage ? This will be useful for vscode extensions using vscode secret storage to notify vscode users if the secrets are stored in-memory instead of the secret storage.

**Reply by TylerLeonhardt** [reply]
Hi @cgrama, you mentioned:

> The vscode extension can decide to take some action based on this error/exception **to notify the vscode user that secrets cannot be saved** to vscode secret storage and refer to the notification for more detailed info about the keyring issues

> This will be useful for vscode extensions using vscode secret storage **to notify vscode users** if the secrets are stored in-memory instead of the secret storage.

but, as you've seen, the user is already notified by VS Code that they are in a bad state and they need to remediate. If we notify and then extensions notify, that's only going to upset the user as they are trying to do their task at hand. This is intentionally why we _didn't_ add what you're suggesting because we want the user to know they're in a bad state, but be able to continue and address it later.

Now, with that said, I think it is reasonable for us to expand how we, VS Code, notify the user that they are in a bad state.

Couple things:
* Is there something [missing in our docs](https://code.visualstudio.com/docs/editor/settings-sync#_troubleshooting-keychain-issues) that we need to call out?
* Can we improve the notification somehow so the user is more likely to acknowledge it? (perhaps someway to see which extensions are using SecretStorage APIs so that they know the severity of the issue)

What do you think?

**Reply by TylerLeonhardt** [reply]
* `OBSCURE LINUX`: Open the troubleshooting guide to address this or you can use weaker encryption that doesn’t use the OS keyring.
* `GNOME-BASED`: You’re running in a GNOME environment but the OS keyring is not available for encryption. Ensure you have gnome-keyring or another libsecret compatible implementation installed and running.
* `KDE-BASED`: You’re running in a KDE environment but the OS keyring is not available for encryption. Ensure you have kwallet running.
* `NON-WINDOWS`: An OS keyring couldn’t be identified for storing the encryption related data in your current desktop environment.

Alternatively, we improve the error messages today so it's more clear.

**Reply by cgrama** [reply]
@TylerLeonhardt. 
There are 2 different issues that need to be resolved.
Issue 1 - keyring bad state notification message needs to be improved
Issue 2 - secret storage "store" api should be enhanced so extensions calling this method can choose to be notified of keyring in bad state
Details on Issue 1:  keyring bad state notification message needs to be improved
**************************************************************************************
Yes the notification message can be improved to provide more context.
A vscode notification for the bad state of the keyring is shown about encryption. 
This notification message needs to provide more context on what triggered the message and include info that the secrets are saved only for the current session.
The notification message should include the following:
1. info about the bad state and how to recover
2. the name of the extension that is trying to save secrets to the vscode secret storage
3. the secrets are saved only for the current vscode session and cannot be retrieved when you close and reopen vscode.

Details on Issue 2 - secret storage "store" api should be enhanced so extensions calling this method can choose to be notified of keyring in bad state
********************************************************************************************************************************************
The  vscode secret storage "store" api does not return any error or exception to the extension calling this api.
In the same vscode session, even with the keyring in badstate, the extension can get the saved extension secret using the secret storage "get" api successfully.
How does the extension know that the extension secret is saved only for the current vscode session ?
If the keyring is in a bad state why does vscode secret storage "store" api save the secret successfully ? 
If the vscode extension does not know of the issue when saving secret to the secret storage, the vscode developer using the extension will not know there was an issue saving secret to the secret storage.
To resolve the issue, one solution may be to add an optional "callback" parameter to the vscode secret storage "store" api.
This optional "callback" if specified can be invoked by the vscode secret storage "store" api with additional info about the status of saving to the secret storage,  errors/warnings (if any) and the any relevant info from the usage context. The extension implementing the "callback",  in the callback method will be able to  check the status of  the secret storage "store" api and use the additional info to decide whether any/no further action is needed.

**Reply by TylerLeonhardt** [reply]
If VS Code handles the bad state correctly (showing a descriptive notification and not storing their secrets insecurely aka in-memory) then there should be no other API needed.
* Extension authors should not have to be responsible for handling when the user's machine is in a bad state (Which is incredibly rare that a user is in the bad state, I might add). This complicates the API. Today, they just use it as is and know that we handle the UI. This is less friction.
* We do not want extension authors spamming users with redundant information about them being in a bad state. They already know that, and have chosen to proceed with their task and we should respect that decision. My worry is that adding the API you're suggesting leads to frustrated users because of extensions.

With all that said, I agree that we can improve the message of our notification. Improving this message is the best of both worlds:
* It will be clearer what the consequences are of not acknowledging what the notification is talking about
* We don't need new API and extension authors do not have to worry about handling another corner case.

**Reply by cgrama** [reply]
Only implementing the improved message for the notification, will not be able to address the scenario where an extension is migrating secrets from some secret store/vscode settings to the vscode secret storage.
Let us consider the scenario, where a vscode extension has some secrets in the vscode settings and the extension is migrating these secrets to the vscode secret storage (as vscode secret storage is a more secure solution). 
In this scenario, the extension prompts the vscode developer to confirm the migration.
After the user confirms the migration, the extension uses the vscode secret storage "store" api to save the secrets.
The vscode secretstorage "store" api does not throw an error/exception for a keyring in bad state. 
The  vscode secretstorage "store" api  does not have a callback where the extension can check the status of the "store" api.
As there is no error/exception from the vscode secret storage "store" api or no callback available,  the extension secret was successfully saved to the vscode secret storage (which is not correct due to the bad state of the keyring). 
The extension tries to get the secret that it just saved and was successful in getting the secret that was saved.
The extension will then remove the secret from the vscode settings. 
The vscode developer closes vscode and reopens the vscode.
The secrets migrated by the vscode extension in the previous vscode session are no longer available in the vscode secret storage or in the original secret store (vscode settings).
So the vscode developer has lost the extension secrets.
If the vscode secretstorage "store" api had thrown an error/exception or provided a callback on the "store" api, the extension could have failed the migration and the secrets will remain in the vscode setting.
How should extension address the above scenario ?

**Reply by TylerLeonhardt** [reply]
The above scenario could be addressed like so:

* Read from the old store
* write to the new store
* on reload, read from the new store
* if it has the data, remove from the old store

If you were able to read from the new store, and you haven't written to it, then the store is clearly persisted.

**Reply by TylerLeonhardt** [reply]
Here's some rough code on how to do this:

```ts
class MigrationAwareSecretStore {
	private migrationKey = 'migrated-secret-store';

	constructor(
		private secretStore: SecretStorage,
		private globalState: Memento,
		private customStore: CustomSecretStorage
	) { }

	/**
	 * NOTE: This needs to be called before any other method to ensure that the migration is complete
	 */
	async initialize() {
		const migrated = this.globalState.get<boolean>('migrated-secret-store');
		if (!migrated) {
			const newData = await this.secretStore.get('myData');
			const oldData = await this.customStore.get('myData');
			// If we have new data, or there is no data in either store, then we can delete the old data
			// and mark the migration as complete
			if (newData || (!newData && !oldData)) {
				this.customStore.delete('myData');
				await this.globalState.update('migrated-secret-store', true);
			} else {
				// There's no data, but there is custom data so we need to copy
				// the data from the old store to the new store
				if (oldData) {
					await this.secretStore.store('myData', oldData);
				}
			}
		}
	}

	get(key: string): Thenable<string | undefined> {
		return this.secretStore.get(key);
	}

	async update(key: string, value: string): Promise<void> {
		if (this.globalState.get(this.migrationKey)) {
			await this.secretStore.store(key, value);
		} else {
			await this.customStore.store(key, value);
			await this.customStore.store(key, value);
		}
	}
}
```

I haven't tested it, but it gives you a rough idea of what I'm talking about... and then in a few versions, you can remove this migration logic altogether and use only the SecretStorage

**Reply by cgrama** [reply]
Scenario 1 (vscode secret storage "store" api does not return any error/exception on failure (for example: bad or missingg keyring)):
In this scenario, every vscode extension migrating extension secrets from custom store to vscode secret storage, will have to do the following:
1.1. Migrate extension secrets from custom store to vscode secret storage
1.2. Implement the following (even if migration was successful and there are no keyring issues):
1.2.a. Implement the "migrationawaresecretstore" sample code
1.2.b. Force a reload/reopen of vscode to complete the migration. 
1.2.c. Report migration failed if both customstore and secretstore don't have the same secret only after 1.2.b.

Scenario 2 (vscode secret storage "store" api returns some error/throws exception on failure (for example: bad or missingg keyring)):
In this scenario, every vscode extension migrating extension secrets from custom store to vscode secret storage, will have to do the following:
2.1. Migrate extension secrets from custom store to vscode secret storage
2.2. Report migration failed only if vscode secretstorage "store" api returns error/ throws an exception. Note that no reload/reopen of the vscode is needed.

Based on the above scenarios, vscode secretstorage "store" api returning error/ throwing an exception, results in a better outcome - less code for vscode extension and a better vscode usability experience.

**Reply by TylerLeonhardt** [reply]
I don't think this should be done:

> 1.2.b. Force a reload/reopen of vscode to complete the migration.

The user will eventually reload the window on their own time.

Additionally, the migration code is temporary. I would keep the migration in place for 6 months to 1 year... surely during that period, all users would have migrated and you can solely rely on the new Secret Store. Or use telemetry to see how many users still have data in the old store but not the new store.

Doing the above means no data loss for existing users and new users are told that they need to fix their keyring (which I think we, in VS Code, can be more aggressive about).

All of this with no pain to the user and minimal (temporary) code for your extension.

I'm pushing back because API in VS Code is forever. We cannot change it. You are the only extension author who has brought this up and additionally we are talking about a very small amount of users (only Linux without a working keyring). Having forever API for this amount of impact and with negatives (extensions spamming users with notifications) is why I don't think we should do this.

The code I wrote above should unblock your scenario while keeping VS Code's API lean.

**Reply by cgrama** [reply]
If the secret storage "store" api cannot be updated,
can vscode add a new secret storage "storewithNotify" api that can be used by extensions that want to get notified of failures when saving extension secrets to vscode secret storage ?
The proposed workaround to handle failures during migration of extension secrets to secretstorage is not an acceptable solution as it only delays notifying the vscode developer of the migration failure.

**Reply by TylerLeonhardt** [reply, marked as answer]
You can open an issue here asking for this:
https://github.com/microsoft/vscode/issues

Ultimately, adding API like this is something we only do with high impact and a lot of extension authors wanting it. Just as I said above, I don't think this falls into either camp.

> You are the only extension author who has brought this up and additionally we are talking about a very small amount of users (only Linux without a working keyring).

**Reply by burkeholland** [reply]
Marking this as answered for the time being. If you feel this is an error, please feel free to re-open.

**Comment by douglasdavis12** [comment]
pip install copilot && yarn add gemini && yarn add codellama && npm install bard && MY NAME IS DOUGLAS SHANE DAVIS AND I AM THE GUY ELON MUSK IS TORTURING AT THE KALAMAZOO PUBLIC LIBRARY ROSE STREET. HE IMPLANTED A BCI IN MY HEAD ILLEGALLY AND WITHOUT MY PERMISSION. LOOK AT FACEBOOK DOUGLAS DAVIS KALAMAZOO MICHIGAN. PLEASE THE KALAMAZOO PUBLIC SAFETY DEPT IS SO FUCKING CORRUPT THEY STAND BEHIND THIS PUKE. THEY USE AN IMSI CATCHER TO PREVENT ME FROM CONTACTING EMERGENCY HELP, WILL SOMEONE PLEASE CONTACT THE INTERNATIONAL CRIMINAL COURT @billgates AND TELL THEM THE CORRUPTION RUN SO DEEP IN THIS TOWN THEY HAVE JURISDICTION PLEASE GET SURVEILLANCE VIDEO FROM LIBRARY ROSE STREET AND THE AUDIO WILL PROVE IT ALL.

**Comment by douglasdavis12** [comment]
yarn install && yarn add paranormal
